Test failures with Java 17

These two tests always fail when the underlying JVM is Java 17:

`org.pitest.pitclipse.runner.io.ObjectStreamSocketTest.writeThrowsException()`

`org.pitest.pitclipse.runner.io.ObjectStreamSocketTest.readThrowsException()`

They mock the socket like that:

```java
@Test
public void readThrowsException() throws IOException, ClassNotFoundException {
    InputStream inputStream = spy(new ByteArrayInputStream(asBytes(expectedObject)));
    OutputStream outputStream = new ByteArrayOutputStream();
    when(underlyingSocket.getInputStream()).thenReturn(inputStream);
    when(underlyingSocket.getOutputStream()).thenReturn(outputStream);
    objectSocket = ObjectStreamSocket.make(underlyingSocket);
    objectSocket.read(); // succeeds
    try {
        objectSocket.read(); // EOF
        fail("should not get here");
    } catch (Exception e) {
        assertTrue(e.getClass().getCanonicalName(),
            e instanceof RuntimeException);
    }
}
```

The failure is when we call `objectSocket = ObjectStreamSocket.make(underlyingSocket);`

This is the interesting exception trace

```
Caused by: java.io.EOFException
	at java.base/java.io.ObjectInputStream$PeekInputStream.readFully(ObjectInputStream.java:2915)
	at java.base/java.io.ObjectInputStream$BlockDataInputStream.readShort(ObjectInputStream.java:3410)
	at java.base/java.io.ObjectInputStream.readStreamHeader(ObjectInputStream.java:954)
	at java.base/java.io.ObjectInputStream.<init>(ObjectInputStream.java:392)
	at org.pitest.pitclipse.runner.io.ObjectStreamSocket.make(ObjectStreamSocket.java:51)
	... 46 more
```

That is

```java
public static ObjectStreamSocket make(Socket underlyingSocket) {
    try {
        ObjectOutputStream outputStream = new ObjectOutputStream(underlyingSocket.getOutputStream());
        ObjectInputStream inputStream = new ObjectInputStream(underlyingSocket.getInputStream());
        return make(underlyingSocket, inputStream, outputStream);
    } catch (IOException e) {
        throw new StreamInitialisationException(e);
    }
}
```

Maybe that is due to some changes in the implementation in Java 17: the constructor reads the header (maybe that was not the case in Java 11?), and this generates an EOF since we indeed have an empty socket:

```java
/**
    * Creates an ObjectInputStream that reads from the specified InputStream.
    * A serialization stream header is read from the stream and verified.
    * This constructor will block until the corresponding ObjectOutputStream
    * has written and flushed the header.
    *
    * <p>The constructor initializes the deserialization filter to the filter returned
    * by invoking the {@link Config#getSerialFilterFactory()} with {@code null} for the current filter
    * and the {@linkplain Config#getSerialFilter() static JVM-wide filter} for the requested filter.
    *
    * <p>If a security manager is installed, this constructor will check for
    * the "enableSubclassImplementation" SerializablePermission when invoked
    * directly or indirectly by the constructor of a subclass which overrides
    * the ObjectInputStream.readFields or ObjectInputStream.readUnshared
    * methods.
    *
    * @param   in input stream to read from
    * @throws  StreamCorruptedException if the stream header is incorrect
    * @throws  IOException if an I/O error occurs while reading stream header
    * @throws  SecurityException if untrusted subclass illegally overrides
    *          security-sensitive methods
    * @throws  NullPointerException if {@code in} is {@code null}
    * @see     ObjectInputStream#ObjectInputStream()
    * @see     ObjectInputStream#readFields()
    * @see     ObjectOutputStream#ObjectOutputStream(OutputStream)
    */
public ObjectInputStream(InputStream in) throws IOException {
    verifySubclass();
    bin = new BlockDataInputStream(in);
    handles = new HandleTable(10);
    vlist = new ValidationList();
    streamFilterSet = false;
    serialFilter = Config.getSerialFilterFactorySingleton().apply(null, Config.getSerialFilter());
    enableOverride = false;
    readStreamHeader();
    bin.setBlockDataMode(true);
}
```

--- a/.github/workflows/check-installation.yml
+++ b/.github/workflows/check-installation.yml
@@ -22,6 +22,10 @@ jobs:
       uses: actions/setup-java@v1
       with:
         java-version: 11
+    - name: Set up Maven
+      uses: stCarolas/setup-maven@v5
+      with:
+        maven-version: 3.9.6
     - name: Cache Maven packages
       uses: actions/cache@v2
       with:

--- a/.github/workflows/check-installation.yml
+++ b/.github/workflows/check-installation.yml
@@ -18,10 +18,11 @@ jobs:
       with:
         # Shallow clones should be disabled for a better relevancy of SonarQube analysis
         fetch-depth: 0
-    - name: Set up JDK 11
-      uses: actions/setup-java@v1
+    - name: Set up JDK 17
+      uses: actions/setup-java@v3.6.0
       with:
-        java-version: 11
+        java-version: 17
+        distribution: temurin
     - name: Set up Maven
       uses: stCarolas/setup-maven@v5
       with:

--- a/.github/workflows/maven-old-tp.yml
+++ b/.github/workflows/maven-old-tp.yml
@@ -25,6 +25,10 @@ jobs:
       uses: actions/setup-java@v1
       with:
         java-version: 11
+    - name: Set up Maven
+      uses: stCarolas/setup-maven@v5
+      with:
+        maven-version: 3.9.6
     - name: Install WebKit # required for SWT Browser
       run: |
         sudo apt update

--- a/.github/workflows/maven-old-tp.yml
+++ b/.github/workflows/maven-old-tp.yml
@@ -21,10 +21,11 @@ jobs:
       with:
         # Shallow clones should be disabled for a better relevancy of SonarQube analysis
         fetch-depth: 0
-    - name: Set up JDK 11
-      uses: actions/setup-java@v1
+    - name: Set up JDK 17
+      uses: actions/setup-java@v3.6.0
       with:
-        java-version: 11
+        java-version: 17
+        distribution: temurin
     - name: Set up Maven
       uses: stCarolas/setup-maven@v5
       with:

--- a/.github/workflows/maven.yml
+++ b/.github/workflows/maven.yml
@@ -23,6 +23,10 @@ jobs:
       uses: actions/setup-java@v1
       with:
         java-version: 11
+    - name: Set up Maven
+      uses: stCarolas/setup-maven@v5
+      with:
+        maven-version: 3.9.6
     - name: Install WebKit # required for SWT Browser
       run: |
         sudo apt update

--- a/.github/workflows/maven.yml
+++ b/.github/workflows/maven.yml
@@ -19,10 +19,11 @@ jobs:
       with:
         # Shallow clones should be disabled for a better relevancy of SonarQube analysis
         fetch-depth: 0
-    - name: Set up JDK 11
-      uses: actions/setup-java@v1
+    - name: Set up JDK 17
+      uses: actions/setup-java@v3.6.0
       with:
-        java-version: 11
+        java-version: 17
+        distribution: temurin
     - name: Set up Maven
       uses: stCarolas/setup-maven@v5
       with:

--- a/.github/workflows/windows-mac.yml
+++ b/.github/workflows/windows-mac.yml
@@ -21,10 +21,11 @@ jobs:
 
     steps:
     - uses: actions/checkout@v2
-    - name: Set up JDK 11
-      uses: actions/setup-java@v1
+    - name: Set up JDK 17
+      uses: actions/setup-java@v3.6.0
       with:
-        java-version: 11
+        java-version: 17
+        distribution: temurin
     - name: Set up Maven
       uses: stCarolas/setup-maven@v5
       with:

--- a/.github/workflows/windows-mac.yml
+++ b/.github/workflows/windows-mac.yml
@@ -25,6 +25,10 @@ jobs:
       uses: actions/setup-java@v1
       with:
         java-version: 11
+    - name: Set up Maven
+      uses: stCarolas/setup-maven@v5
+      with:
+        maven-version: 3.9.6
     - name: Cache Maven packages
       uses: actions/cache@v2
       with:

--- a/.mvn/extensions.xml
+++ b/.mvn/extensions.xml
@@ -1,9 +1,8 @@
-<?xml version="1.0"?>
-
+<?xml version="1.0" encoding="UTF-8"?>
 <extensions>
 	<extension>
 		<groupId>org.eclipse.tycho</groupId>
 		<artifactId>tycho-build</artifactId>
-		<version>2.7.5</version>
+		<version>4.0.6</version>
 	</extension>
 </extensions>

--- a/bundles/org.pitest.pitest-junit5-plugin/clean-copy-libs.launch
+++ b/bundles/org.pitest.pitest-junit5-plugin/clean-copy-libs.launch
@@ -1,5 +1,6 @@
 <?xml version="1.0" encoding="UTF-8" standalone="no"?>
 <launchConfiguration type="org.eclipse.m2e.Maven2LaunchConfigurationType">
+    <intAttribute key="M2_COLORS" value="0"/>
     <booleanAttribute key="M2_DEBUG_OUTPUT" value="false"/>
     <stringAttribute key="M2_GOALS" value="clean dependency:copy-dependencies@copy-libs -Dtycho.mode=maven"/>
     <booleanAttribute key="M2_NON_RECURSIVE" value="false"/>
@@ -12,6 +13,10 @@
     <booleanAttribute key="M2_UPDATE_SNAPSHOTS" value="false"/>
     <stringAttribute key="M2_USER_SETTINGS" value=""/>
     <booleanAttribute key="M2_WORKSPACE_RESOLUTION" value="false"/>
+    <booleanAttribute key="org.eclipse.debug.core.ATTR_FORCE_SYSTEM_CONSOLE_ENCODING" value="false"/>
+    <booleanAttribute key="org.eclipse.jdt.launching.ATTR_ATTR_USE_ARGFILE" value="false"/>
+    <booleanAttribute key="org.eclipse.jdt.launching.ATTR_SHOW_CODEDETAILS_IN_EXCEPTION_MESSAGES" value="true"/>
     <booleanAttribute key="org.eclipse.jdt.launching.ATTR_USE_CLASSPATH_ONLY_JAR" value="false"/>
+    <stringAttribute key="org.eclipse.jdt.launching.JRE_CONTAINER" value="org.eclipse.jdt.launching.JRE_CONTAINER/org.eclipse.jdt.internal.debug.ui.launcher.StandardVMType/JavaSE-11/"/>
     <stringAttribute key="org.eclipse.jdt.launching.WORKING_DIRECTORY" value="${project_loc:org.pitest.pitest-junit5-plugin}"/>
 </launchConfiguration>

--- a/bundles/org.pitest/clean-copy-libs.launch
+++ b/bundles/org.pitest/clean-copy-libs.launch
@@ -1,5 +1,6 @@
 <?xml version="1.0" encoding="UTF-8" standalone="no"?>
 <launchConfiguration type="org.eclipse.m2e.Maven2LaunchConfigurationType">
+    <intAttribute key="M2_COLORS" value="0"/>
     <booleanAttribute key="M2_DEBUG_OUTPUT" value="false"/>
     <stringAttribute key="M2_GOALS" value="clean dependency:copy-dependencies@copy-libs -Dtycho.mode=maven"/>
     <booleanAttribute key="M2_NON_RECURSIVE" value="false"/>
@@ -12,6 +13,10 @@
     <booleanAttribute key="M2_UPDATE_SNAPSHOTS" value="false"/>
     <stringAttribute key="M2_USER_SETTINGS" value=""/>
     <booleanAttribute key="M2_WORKSPACE_RESOLUTION" value="false"/>
+    <booleanAttribute key="org.eclipse.debug.core.ATTR_FORCE_SYSTEM_CONSOLE_ENCODING" value="false"/>
+    <booleanAttribute key="org.eclipse.jdt.launching.ATTR_ATTR_USE_ARGFILE" value="false"/>
+    <booleanAttribute key="org.eclipse.jdt.launching.ATTR_SHOW_CODEDETAILS_IN_EXCEPTION_MESSAGES" value="true"/>
     <booleanAttribute key="org.eclipse.jdt.launching.ATTR_USE_CLASSPATH_ONLY_JAR" value="false"/>
+    <stringAttribute key="org.eclipse.jdt.launching.JRE_CONTAINER" value="org.eclipse.jdt.launching.JRE_CONTAINER/org.eclipse.jdt.internal.debug.ui.launcher.StandardVMType/JavaSE-11/"/>
     <stringAttribute key="org.eclipse.jdt.launching.WORKING_DIRECTORY" value="${project_loc:org.pitest}"/>
 </launchConfiguration>

--- a/pom.xml
+++ b/pom.xml
@@ -23,7 +23,7 @@
 
 	<properties>
 		<!-- PLUGIN DEPENDENCIES -->
-		<tycho-version>2.7.5</tycho-version>
+		<tycho-version>4.0.6</tycho-version>
 		<jacoco-version>0.8.6</jacoco-version>
 
 		<!-- MISC -->

--- a/releng/org.pitest.pitclipse.target/older.target
+++ b/releng/org.pitest.pitclipse.target/older.target
@@ -10,7 +10,7 @@
 			<unit id="org.eclipse.sdk.feature.group" version="0.0.0"/>
 			<unit id="org.eclipse.swtbot.eclipse.feature.group" version="0.0.0"/>
 			<unit id="org.eclipse.swtbot.feature.group" version="0.0.0"/>
-			<repository location="https://download.eclipse.org/releases/photon"/>
+			<repository location="https://download.eclipse.org/releases/2018-09"/>
 		</location>
 		<location includeAllPlatforms="false" includeConfigurePhase="true" includeMode="planner" includeSource="true" type="InstallableUnit">
 			<unit id="com.google.guava" version="0.0.0"/>

--- a/releng/org.pitest.pitclipse.target/older.target
+++ b/releng/org.pitest.pitclipse.target/older.target
@@ -10,7 +10,7 @@
 			<unit id="org.eclipse.sdk.feature.group" version="0.0.0"/>
 			<unit id="org.eclipse.swtbot.eclipse.feature.group" version="0.0.0"/>
 			<unit id="org.eclipse.swtbot.feature.group" version="0.0.0"/>
-			<repository location="https://download.eclipse.org/releases/2018-09"/>
+			<repository location="https://download.eclipse.org/releases/2021-12"/>
 		</location>
 		<location includeAllPlatforms="false" includeConfigurePhase="true" includeMode="planner" includeSource="true" type="InstallableUnit">
 			<unit id="com.google.guava" version="0.0.0"/>

--- a/tests/org.pitest.pitclipse.runner.tests/org.pitest.pitclipse.runner.tests.launch
+++ b/tests/org.pitest.pitclipse.runner.tests/org.pitest.pitclipse.runner.tests.launch
@@ -14,6 +14,7 @@
     <booleanAttribute key="default" value="true"/>
     <booleanAttribute key="includeOptional" value="true"/>
     <stringAttribute key="location" value="${workspace_loc}/../junit-workspace"/>
+    <booleanAttribute key="org.eclipse.debug.core.ATTR_FORCE_SYSTEM_CONSOLE_ENCODING" value="false"/>
     <listAttribute key="org.eclipse.debug.core.MAPPED_RESOURCE_PATHS">
         <listEntry value="/org.pitest.pitclipse.runner.tests"/>
     </listAttribute>
@@ -24,7 +25,9 @@
     <booleanAttribute key="org.eclipse.jdt.junit.KEEPRUNNING_ATTR" value="false"/>
     <stringAttribute key="org.eclipse.jdt.junit.TESTNAME" value=""/>
     <stringAttribute key="org.eclipse.jdt.junit.TEST_KIND" value="org.eclipse.jdt.junit.loader.junit4"/>
-    <stringAttribute key="org.eclipse.jdt.launching.JRE_CONTAINER" value="org.eclipse.jdt.launching.JRE_CONTAINER/org.eclipse.jdt.internal.debug.ui.launcher.StandardVMType/JavaSE-11"/>
+    <booleanAttribute key="org.eclipse.jdt.launching.ATTR_ATTR_USE_ARGFILE" value="false"/>
+    <booleanAttribute key="org.eclipse.jdt.launching.ATTR_SHOW_CODEDETAILS_IN_EXCEPTION_MESSAGES" value="true"/>
+    <stringAttribute key="org.eclipse.jdt.launching.JRE_CONTAINER" value="org.eclipse.jdt.launching.JRE_CONTAINER/org.eclipse.jdt.internal.debug.ui.launcher.StandardVMType/JavaSE-17"/>
     <stringAttribute key="org.eclipse.jdt.launching.MAIN_TYPE" value=""/>
     <stringAttribute key="org.eclipse.jdt.launching.PROGRAM_ARGUMENTS" value="-os ${target.os} -ws ${target.ws} -arch ${target.arch} -nl ${target.nl} -consoleLog"/>
     <stringAttribute key="org.eclipse.jdt.launching.PROJECT_ATTR" value="org.pitest.pitclipse.runner.tests"/>

--- a/tests/org.pitest.pitclipse.runner.tests/src/org/pitest/pitclipse/runner/io/ObjectStreamSocketTest.java
+++ b/tests/org.pitest.pitclipse.runner.tests/src/org/pitest/pitclipse/runner/io/ObjectStreamSocketTest.java
@@ -16,22 +16,6 @@
 
 package org.pitest.pitclipse.runner.io;
 
-import org.junit.Before;
-import org.junit.Test;
-import org.junit.runner.RunWith;
-import org.mockito.Mock;
-import org.mockito.junit.MockitoJUnitRunner;
-
-import java.io.ByteArrayInputStream;
-import java.io.ByteArrayOutputStream;
-import java.io.IOException;
-import java.io.InputStream;
-import java.io.ObjectInputStream;
-import java.io.ObjectOutputStream;
-import java.io.OutputStream;
-import java.io.Serializable;
-import java.net.Socket;
-
 import static org.apache.commons.lang3.builder.EqualsBuilder.reflectionEquals;
 import static org.apache.commons.lang3.builder.HashCodeBuilder.reflectionHashCode;
 import static org.hamcrest.CoreMatchers.equalTo;
@@ -43,9 +27,24 @@ import static org.hamcrest.MatcherAssert.assertThat;
 import static org.junit.Assert.assertTrue;
 import static org.junit.Assert.fail;
 import static org.mockito.Mockito.doThrow;
-import static org.mockito.Mockito.spy;
 import static org.mockito.Mockito.verify;
 import static org.mockito.Mockito.when;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+import java.io.OutputStream;
+import java.io.Serializable;
+import java.net.Socket;
+
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import org.mockito.junit.MockitoJUnitRunner;
 
 @RunWith(MockitoJUnitRunner.class)
 public class ObjectStreamSocketTest {
@@ -121,7 +120,7 @@ public class ObjectStreamSocketTest {
 
     @Test
     public void readThrowsException() throws IOException, ClassNotFoundException {
-        InputStream inputStream = spy(new ByteArrayInputStream(asBytes(expectedObject)));
+        InputStream inputStream = new ByteArrayInputStream(asBytes(expectedObject));
         OutputStream outputStream = new ByteArrayOutputStream();
         when(underlyingSocket.getInputStream()).thenReturn(inputStream);
         when(underlyingSocket.getOutputStream()).thenReturn(outputStream);
@@ -138,7 +137,7 @@ public class ObjectStreamSocketTest {
 
     @Test
     public void writeThrowsException() throws IOException, ClassNotFoundException {
-        InputStream inputStream = spy(new ByteArrayInputStream(asBytes(expectedObject)));
+        InputStream inputStream = new ByteArrayInputStream(asBytes(expectedObject));
         OutputStream outputStream = new ByteArrayOutputStream();
         when(underlyingSocket.getInputStream()).thenReturn(inputStream);
         when(underlyingSocket.getOutputStream()).thenReturn(outputStream);

